src/python/gem5/resources/resource.py

```diff
@@ -750,6 +807,52 @@ def get_input_groups(self) -> Set[str]:
         }
 
 
+class ShadowResource(AbstractResource):
+    """A special resource class which delays the `obtain_resource` call. It is,
+    in a sense, half constructed. Only when a function or attribute is called
+    which is is neither `get_id` or `get_resource_version` does this class
+    fully construct itself by calling the `obtain_resource_call` partial
+    function.
+    """
+
+    def __init__(
+        self,
+        id: str,
+        resource_version: str,
+        obtain_resource_call: partial,
+    ):
+        super().__init__(
+            id=id,
+            resource_version=resource_version,
+        )
+        self._workload: Optional[AbstractResource] = None
+        self._obtain_resource_call = obtain_resource_call
+
+    def __getattr__(self, attr):
+        """if getting the id or resource version, we keep the object in the
+        "shdow state" where the `obtain_resource` function has not been called.
+        When more information is needed by calling another attribute, we call
+        the `obtain_resource` function and store the result in the `_workload`.
+        """
+        if attr in {"get_id", "get_resource_version"}:
+            return getattr(super(), attr)
+        if not self._workload:
+            self._workload = self._obtain_resource_call()
+        return getattr(self._workload, attr)
+
+    def set_parameter(self, parameter: str, value: Any) -> None:
+        """
+        Used to set or override a workload parameter
+
+        :param parameter: The parameter of the function to set.
+        :param value: The value to set to the parameter.
+        """
+        self._params[parameter] = value
+
+    def get_category_name(cls) -> str:
+        return "WorkloadResource"
```

Comment:
if this is only used for lazy workloads, then instead of `ShadowResource` this class should be something like `Lazy/ShadowWorkloadResource`

Reply:
I didn't mean to leave in these functions. Removing.